Better README. (#91)
ocaml-multicore/saturn: README.md

> [API reference](https://ocaml-multicore.github.io/saturn/)

Reviewer: ```suggestion
[API Reference](https://ocaml-multicore.github.io/saturn/)
```

<!--
```ocaml
# #thread
```

> -->

Reviewer 2: This comment can be removed, I guess.

Reviewer 3: It depends.  If you need threads then on some platforms the `#thread` to enable threads has to be first.  Otherwise things don't work.

Reviewer 2: I see. Makes sense to keep it.

> # Saturn — parallelism-safe data structures for Multicore OCaml

Reviewer: ```suggestion
# Saturn — Parallelism-Safe Data Structures for Multicore OCaml
```

This repository is a collection of parallelism-safe data structures for OCaml 5.
They are contained in two packages:

> - Saturn that includes all data structures and should be used by default if you

Reviewer 2: Maybe it's useful to make it clear that Saturn exposes both lockfree and other parallelism-safe data structures.

  just want parallelism-safe data structures;
- Saturn_lockfree that includes only lock-free data structures.

> It aims to provide an industria-strenght, well-tested (and possibly

Reviewer 2: ```suggestion
It aims to provide an industrial-strength, well-tested (and possibly
```

> model-checked and verified in the future), well documented and maintained

Reviewer: ```suggestion
model-checked and verified in the future), well documented, and maintained
```

> parallelism-safe data structure library. We want to make it easier for multicore

Reviewer: ```suggestion
parallelism-safe data structure library. We want to make it easier for Multicore
```

is distributed under the
[ISC license](https://github.com/ocaml-multicore/saturn/blob/main/LICENSE.md).

[![OCaml-CI Build Status](https://img.shields.io/endpoint?url=https%3A%2F%2Fci.ocamllabs.io%2Fbadge%2Focaml-multicore%2Fsaturn%2Fmain&logo=ocaml&style=flat-square)](https://ci.ocamllabs.io/github/ocaml-multicore/saturn)
[![GitHub release (latest by date)](https://img.shields.io/github/v/release/ocaml-multicore/saturn?style=flat-square&color=09aa89)](https://github.com/ocaml-multicore/saturn/releases/latest)
[![docs](https://img.shields.io/badge/doc-online-blue.svg?style=flat-square)](https://ocaml-multicore.github.io/saturn/)

# Contents

- [Installation](#installation)
- [Introduction](#introduction)
  - [Provided data structures](#provided-data-structures)
  - [Motivation](#motivation)
- [Usage](#usage)
  - [Data structures with domain roles](#data-structures-with-domain-roles)
  - [About composability](#about-composability)
- [Properties and testing](#properties-and-testing)
  - [Progress properties](#progress-properties)
  - [Tests](#tests)
- [Benchmarks](#benchmarks)

# Installation

## Getting OCaml 5.0

> You'll need OCaml 5.0.0 or later. Note that `saturn` also works with OCaml 4.14

Reviewer: Is this `saturn` a command or should it be Saturn, the name of the data structures?

> but only for compatible reasons as there is no need for parallislim-safe data

Reviewer 2: ```suggestion
but only for compatibility reasons as there is no need for parallislim-safe data
```

Reviewer: Also a comma before as "...reasons, as there is no need..."

Reviewer 4: Agreed!

| Michael-Scott queue             | [`Queue`](https://ocaml-multicore.github.io/saturn/saturn_lockfree/Lockfree/Queue/index.html) (same)                                                                                                    | A classic multi-producer multi-consumer queue, robust and flexible. Recommended starting point when needing a FIFO structure.                                                                                                                                                  | [Simple, Fast, and Practical Non-Blocking and Blocking Concurrent Queue Algorithms](https://www.cs.rochester.edu/~scott/papers/1996_PODC_queues.pdf)                                                         |

> | Chase-Lev Work-Stealing Dequeue | [`Work_stealing_deque`](https://ocaml-multicore.github.io/saturn/saturn_lockfree/Lockfree/Work_stealing_deque/index.html) (same)                                                                        | Single-producer, multi-consumer dynamic-size double-ended queue (deque). Ideal for throughput-focused scheduling using per-core work distribution. Note, `pop` and `steal` follow different ordering (respectively LIFO and FIFO) and have different linearization contraints. | [Dynamic circular work-stealing deque](https://dl.acm.org/doi/10.1145/1073970.1073974) and [Correct and efficient work-stealing for weak memory models](https://dl.acm.org/doi/abs/10.1145/2442516.2442524)) |

Reviewer: Perhaps commas between "...multi-user, dynamic-size, double-ended queue (deque)."

> | Chase-Lev Work-Stealing Dequeue | [`Work_stealing_deque`](https://ocaml-multicore.github.io/saturn/saturn_lockfree/Lockfree/Work_stealing_deque/index.html) (same)                                                                        | Single-producer, multi-consumer dynamic-size double-ended queue (deque). Ideal for throughput-focused scheduling using per-core work distribution. Note, `pop` and `steal` follow different ordering (respectively LIFO and FIFO) and have different linearization contraints. | [Dynamic circular work-stealing deque](https://dl.acm.org/doi/10.1145/1073970.1073974) and [Correct and efficient work-stealing for weak memory models](https://dl.acm.org/doi/abs/10.1145/2442516.2442524)) |

Author: Thanks for your reviews @christinerose  @ILeandersson !

For this particular comment,  we need to distinguish between users that take (consumers) and users that add (producers). Producer/consumer is a commonly used vocabulary in computer science :)

| SPSC Queue                      | [`Single_prod_single_`<br>`cons_queue`](https://ocaml-multicore.github.io/saturn/saturn_lockfree/Lockfree/Single_prod_single_cons_queue/index.html) (same)                                              | Simple single-producer single-consumer fixed-size queue. Thread-safe as long as at most one thread acts as producer and at most one as consumer at any single point in time.                                                                                                   |                                                                                                                                                                                                              |

> | MPMC bounded relaxed queue      | [`Relaxed_queue`](https://ocaml-multicore.github.io/saturn/saturn/Saturn/Relaxed_queue/index.html) ([same](https://ocaml-multicore.github.io/saturn/saturn_lockfree/Lockfree/Relaxed_queue/index.html)) | Multi-producer, multi-consumer, fixed-size relaxed queue. Optimised for high number of threads. Not strictly FIFO. Note, it exposes two interfaces: a lockfree and a non-lockfree (albeit more practical) one. See the mli for details.                                        |                                                                                                                                                                                                              |

Reviewer: ```suggestion
| MPMC bounded relaxed queue      | [`Relaxed_queue`](https://ocaml-multicore.github.io/saturn/saturn/Saturn/Relaxed_queue/index.html) ([same](https://ocaml-multicore.github.io/saturn/saturn_lockfree/Lockfree/Relaxed_queue/index.html)) | Multi-producer, multi-consumer, fixed-size relaxed queue. Optimised for high number of threads. Not strictly FIFO. Note, it exposes two interfaces: a lockfree and a non-lockfree (albeit more practical) one. See the `.mli` for details.                                        |                                                                                                                                                                                                              |
```

Reviewer: Also, should the MPMC Bounded Relaxed Queue be capitalised? It looks like some of the other ones are title case. 

If so, Michael-Scott Queue and Treiber Stack should also be capitalised (both are correct, depending on how you want to format it, but this is if you want consistency in formatting. Most people won't notice, so it's up to you!)

| MPSC Queue                      | [`Single_consumer_queue`](https://ocaml-multicore.github.io/saturn/saturn_lockfree/Lockfree/Single_consumer_queue/index.html) (same)                                                                    | A multi-producer, single-consumer, thread-safe queue without support for cancellation. This makes a gooddata structure for a scheduler's run queue. It is used in [Eio](https://github.com/ocaml-multicore/eio).                                                               | It is a single consumer version of the queue described in [Implementing lock-free queues](https://people.cs.pitt.edu/~jacklange/teaching/cs2510-f12/papers/implementing_lock_free.pdf).                      |

## Motivation

The following part is a beginner-friendly example to explain why we need data
structures specifically designed for multicore programming.

> Let's write a classic mutable data structures : a queue. We are going to do a

Reviewer: ```suggestion
Let's write a classic mutable data structure : a queue. We are going to do a
```

basic implementation the way it may be done for sequential use and show why it
is not working well with multiple domains.

```ocaml
type queue = int list ref

let create () : queue = ref []

let push q a = q := a :: !q
```

What happens if we try to use this queue with multiple domains ? First, let's
define the work we want a single domain to do: each domain will push 10 times
its `id` in the queue.

```ocaml
let work id q = for i = 0 to 9 do push q id done
```

Then let's define our test : it spawns 2 domains that each executes `work` in
parallel. `test` returns the content of the queue as well as its length, so we
can easily see if it contains 20 elements as expected.

```ocaml
 let test () =
   let q = create () in
   let domainA = Domain.spawn (fun () -> work 1 q) in
   let domainB = Domain.spawn (fun () -> work 2 q) in
   Domain.join domainA;
   Domain.join domainB;
   (List.length !q, !q)
```

Let's try it :

```ocaml
# test ()
- : int * int list =
(20, [2; 2; 2; 2; 2; 2; 2; 2; 2; 2; 1; 1; 1; 1; 1; 1; 1; 1; 1; 1])
```

> Everything seems fine, right ? Except, it is not running in parallel as we can

Reviewer: Perhaps a comma before "as"?

> Everything seems fine, right ? Except, it is not running in parallel as we can

Reviewer 4: I'm neutral on this :)

We use a basic
[barrier implementation](https://github.com/ocaml-multicore/saturn/blob/main/test/barrier/barrier.mli)
to do that. Each domain will now wait for the other to reach the barrier before
beginning pushing.

```ocaml
let work_par id barrier q =
  Barrier.await barrier;
  for i = 0 to 9 do
    push q id
  done
```

The `test` function is now:

```ocaml
let test_par () =
  let barrier = Barrier.create 2 in
  let q = create () in
  let domainA = Domain.spawn (fun () -> work_par 1 barrier q) in
  let domainB = Domain.spawn (fun () -> work_par 2 barrier q) in
  Domain.join domainA;
  Domain.join domainB;
  (List.length !q, !q)
```

Let's run it:

```ocaml
# test_par ();;
- : int * int list =
(18, [2; 1; 2; 1; 2; 1; 2; 1; 2; 1; 2; 2; 1; 1; 1; 1; 2; 1])
```

Now, the `1` and the `2` are interleaved: domains are running in parallel. The
resulting queue however contains only `18` elements whereas `20` were pushed.
This is because we not only have a
[race condition](https://en.wikipedia.org/wiki/Race_condition) here but also
because `push` is a non-atomic operation. It requires to first read the content
of the queue (`!q`) then write in it (`q := ...`). So, for example, when two

> domains try to push in parallel in an empty queue the following sequence can

Reviewer 4: ```suggestion
domains try to push in parallel into an empty queue the following sequence can
```

happen :

> - domain A read the queue : it is empty
> - domain B read the queue : it is still empty

Reviewer: ```suggestion
- domain A reads the queue : it is empty
- domain B reads the queue : it is still empty
```

- domain A pushes `1` on the empty queue it has read before

> - domain B pushes `2` on the empty queue it has read before This sequence

Reviewer: ```suggestion
- domain B pushes `2` on the empty queue it has read before. This sequence
```

>   results in a queue containing only one element of value `2`: the element

Reviewer: ```suggestion
  results in a queue containing only one element of value `2`. The element
```

  pushed by A is lost because B did not see it.

This is a very common issue in parallel programming that can be avoided or

> bypassed in multiple ways, mainly:

Reviewer 4: ```suggestion
bypassed in multiple ways, mainly by:
```

- use [`Atomic`](https://v2.ocaml.org/api/Atomic.html) for mutable variables,
- use a mutual exclusion mechanism like
  [`Mutex`](https://v2.ocaml.org/api/Mutex.html).

However both solutions have their limits. Using mutexes or locks open the way to
deadlock, livelock, priority inversion etc.. and often considerably limit the

> gain in performance of using multiple cores as it limits the part of the code
> effectively running in parallel. In another hand, atomics are -without a complex
> algorithm to combine them- only a solution for a single shared variable.

Reviewer 4: ```suggestion
performance gained by using multiple cores as they limit the parts of the code
effectively running in parallel. On the other hand, atomics are – without a complex
algorithm to combine them – only a solution for a single shared variable.
```

Reviewer: Consider taking the spaces out around the m-dashes (if they'll indeed show up as m-dashes and not hyphens when live) or just put that aside in parentheses.

Reviewer 4: Seems like British English prefers the en dash with spaces either side, (these are en dashes on my keyboard) not sure how this relates to Markdown though.

Let's try to replace references by atomics in our code to prove this point:

```ocaml
type queue = int list Atomic.t

let create () : queue = Atomic.make []

let push (q : queue) a =
  let curr = Atomic.get q in
  let prev = a :: curr in
  Atomic.set q prev
```

We still need to read and write to fulfill the whole push operation.

```ocaml
# test ();;
- : int * int list = (15, [1; 1; 1; 1; 1; 2; 1; 1; 1; 2; 2; 2; 2; 1; 2])
```

and as expected it is not working. The interleaving scenario describes
previously can still happen. Note that, though it is not observable here, this

> is still better then the previous implementation as we are now race-free (see

Reviewer: Perhaps a comma before "as"

> is still better then the previous implementation as we are now race-free (see

Reviewer 4: Agreed!

[here](#a-note-about-races-in-ocaml) for a quick note about races in OCaml).

As a matter of fact, writting a queue with `push` and `pop` functions that are

> both atomic and parallelism-safe is not as easy as it mays

Reviewer: Should this be "are not" instead of "is not"? I get confused with these sometimes when there are compound subjects in this way.

> both atomic and parallelism-safe is not as easy as it mays

Reviewer 4: I read it as 'is not' since the 'is' refers to 'writing a (singular) queue' ?

> With `Saturn`,

Reviewer: Is this `Saturn` a command or is it Saturn the name of the thing itself?

```ocaml
let work_saturn id barrier q () =
  Barrier.await barrier;
  for i = 0 to 9 do
    Saturn.Queue.push q id
  done

let test_saturn () =
  let barrier = Barrier.create 2 in
  let q = Saturn.Queue.create () in
  let d1 = Domain.spawn (work_saturn 1 barrier q) in
  let d2 = Domain.spawn (work_saturn 2 barrier q) in
  Domain.join d1;
  Domain.join d2;
  let rec pop_all acc =
    match Saturn.Queue.pop q with
    | None -> List.rev acc
    | Some elt -> pop_all (elt :: acc)
  in
  let res = pop_all [] in
  (List.length res, res)
```

Let's run it:

```ocaml
# test_saturn ();;
- : int * int list =
(20, [2; 2; 1; 2; 2; 2; 2; 2; 2; 1; 2; 1; 2; 1; 1; 1; 1; 1; 1; 1])
```

### A note about races in OCaml

Because of the great properties of OCaml 5 memory model (see the
[OCaml manual](https://v2.ocaml.org/manual/parallelism.html#s%3Apar_mm_easy) for

> more details), not a lot can go worst here: at least, not data corruption or
> segmentation fault like it could happen in other languages.

Reviewer: ```suggestion
more details), not a lot can go worst here. At least, data corruption or
segmentation fault won't happen like it can in other languages.
```

Can go "worst" or can go "wrong"?

Reviewer 5: I prefer "not a lot can go wrong".

Reviewer 4: Agree with KC here

# Usage

This part provides notes about how to use the provided data structures or more
exactly, what to not do with it. Two main points are discussed:

> - some data structures have restrictions on one domain can do

Reviewer 2: Perhaps - "Some data structures have restrictions on what operations can be performed in a single domain or a set of domains"

- the currently provided data structures are non-composable

> ## Data structures with domain roles

Reviewer: ```suggestion
## Data Structures With Domain Roles
```

There are several provided data structures that are meant to be used with a
certain domain configuration. These restrictions enable to do optimizations but
that breaks safety properties if not respected. Obviously, these restrictions
are not only written in documentation but also on the name of the data structure
itself. Single consumer queue means only one domain should pop.

Let's take the example of `Single_prod_single_cons_queue`. As suggested by the
name, it should be used with only one domain performing `push` (a producer) and
one domain performing `pop` (a consumer) at the same time. Having two or more

> domains try to perform one of these operations at the same time will break
> safety properties of the `queue` and more likely result in unexpected

Reviewer 4: ```suggestion
domains try to perform one of these operations at the same time will break
the safety properties of the `queue` and more than likely result in unexpected
```

Reviewer: Would it be correct to say: "...more likely result in unexpected
domains try to perform one of these operations simultaneously will break..."

I'm confused.... with what are they trying "one of these operations at the same time?"
If they're trying only one of the operations .... and at the same time (or simultaneously, if you take my suggestion) ... there needs to be a second thing they're trying it with, if that makes sense. 

Perhaps you mean "Having two or more domains simultaneously perform one of these operations..."  
(either `push` or `pop`, right?)

This second way makes it clear that the two or more domains are performing simultaneously rather than performing two operations simultaneously. (At least in my head!) What do you think, @ILeandersson?

Reviewer 4: I read it as two or more domains trying to either push or pop at the same time, not sure I'm correct though

Author: It is meant to say that something can go wrong if there are simultaneously one domain doing a `push` and another domain also doing a `push`

(or same with `pop` instead of `push`)

I have changed the text to : 

```suggestion
Let's take the example of `Single_prod_single_cons_queue`. As suggested by the
name, it should be used with only one domain performing `push` (a producer) and
one domain performing `pop` (a consumer) at the same time. Having two or more
domains simultaneously perform `pop` (or `push`) will break the safety properties
of the `queue` and more likely result in unexpected behaviors.L
```

Is it clearer ?

behaviours.

Let's say we give a bad alias to this queue and misuse it.

```ocaml
module Queue = Saturn.Single_prod_single_cons_queue
```

Each domain is going to try to push 10 times in parallel.

```ocaml
let work id barrier q =
  Barrier.await barrier;
  for i = 0 to 9 do
    Queue.push q id
  done
```

Our `test` function returns the queue after two domains try to simustaneously
push.

```ocaml
let test () =
  let q = Queue.create ~size_exponent:5 in
  let barrier = Barrier.create 2 in
  let d1 = Domain.spawn (fun () -> work 1 barrier q) in
  let d2 = Domain.spawn (fun () -> work 2 barrier q) in
  Domain.join d1;
  Domain.join d2;
  q
```

We can then inspect the queue content by popping it into a list.

```ocaml
let get_content q =
  let rec loop acc =
    match Queue.pop q with
    | None -> acc
    | Some a -> loop (a :: acc)
  in
  loop [] |> List.rev
```

Let's run it :

```ocaml
test () |> get_content;;
- : int list = [2; 1; 1; 1; 1; 1; 1; 1; 1; 1; 2]
```

This run results in a queue with only 11 elements. 9 elements are lost to the
misuse of the single consumer single producer queue.

> ## About composability

Reviewer: ```suggestion
## About Composability
```

TODO

> # Properties and testing

Reviewer: ```suggestion
# Properties and Testing
```

> ## Liveness properties

Reviewer: ```suggestion
## Liveness Properties
```

One of the many difficulties of implementating parallelism-safe data structures
is that in addition to providing the same safety properties as sequental one,
they may also have to observe some _liveness_ properties.

## Tests

Several kind of tests are provided for each data structure:

> - _safety_ : unitary tests and `qcheck` tests: check semantics and expected

Reviewer: ```suggestion
- _safety_ : unitary tests and `qcheck` tests check semantics and expected
```
I'm not sure you need a colon here since "check" works as a verb.

>   behaviors with one and more domains;

Reviewer: ```suggestion
  behaviors with one and more domains
```

Reviewer: semicolon not needed between bullet points, but if you choose to remove the colon as suggested, it makes it a complete sentence, so you could put a period (full stop) there instead.

> - _safety and liveness_ : `STM` tests: check _linearizability_ for two domains

Reviewer: Again with this colon, but if you leave it in, people will still understand.

  (see

>   [multicoretests library](https://github.com/ocaml-multicore/multicoretests));

Reviewer: ```suggestion
  [`multicoretests` library](https://github.com/ocaml-multicore/multicoretests))
```
Maybe in `monospace` if it's the name of a library (??)

> - _liveness_ `dscheck`: checks _non-blocking_ property for as many domains as

Reviewer: If you remove the colon above as suggested, remove this one too.

[API Reference](https://ocaml-multicore.github.io/saturn/)

<!--
```ocaml
# #thread
```
-->

# Saturn — Parallelism-Safe Data Structures for Multicore OCaml

This repository is a collection of parallelism-safe data structures for OCaml 5.
They are contained in two packages:

- Saturn that includes all data structures (including the lock-free ones) and
  should be used by default if you just want parallelism-safe data structures;
- Saturn_lockfree that includes only lock-free data structures.

> It aims to provide an industrial-strenght, well-tested (and possibly

Reviewer 2: ```suggestion
It aims to provide an industrial-strength, well-tested (and possibly
```

| MPMC Bounded Relaxed Queue      | [`Relaxed_queue`](https://ocaml-multicore.github.io/saturn/saturn/Saturn/Relaxed_queue/index.html) ([same](https://ocaml-multicore.github.io/saturn/saturn_lockfree/Lockfree/Relaxed_queue/index.html)) | Multi-producer, multi-consumer, fixed-size relaxed queue. Optimised for high number of threads. Not strictly FIFO. Note, it exposes two interfaces: a lockfree and a non-lockfree (albeit more practical) one. See the `mli` for details.                                        |                                                                                                                                                                                                              |

> | MPSC Queue                      | [`Single_consumer_queue`](https://ocaml-multicore.github.io/saturn/saturn_lockfree/Lockfree/Single_consumer_queue/index.html) (same)                                                                    | A multi-producer, single-consumer, thread-safe queue without support for cancellation. This makes a gooddata structure for a scheduler's run queue. It is used in [Eio](https://github.com/ocaml-multicore/eio).                                                                 | It is a single consumer version of the queue described in [Implementing Lock-Free Queues](https://people.cs.pitt.edu/~jacklange/teaching/cs2510-f12/papers/implementing_lock_free.pdf).                      |

Reviewer 6: typo: gooddata

# test_par ();;
- : int * int list =
(18, [2; 1; 2; 1; 2; 1; 2; 1; 2; 1; 2; 2; 1; 1; 1; 1; 2; 1])
```

Now, the `1` and the `2` are interleaved: domains are running in parallel. The
resulting queue however only contains `18` elements whereas `20` were pushed.
This is because we not only have a
[race condition](https://en.wikipedia.org/wiki/Race_condition) here but also
because `push` is a non-atomic operation. It requires first to read the content
of the queue (`!q`) then to write in it (`q := ...`). So, for example, when two
domains try to push in parallel into an empty queue the following sequence can
happen:

- domain A reads the queue : it is empty
- domain B reads the queue : it is still empty
- domain A pushes `1` on the empty queue it has read before
- domain B pushes `2` on the empty queue it has read before.

This sequence results in a queue containing only one element of value `2`. The
element pushed by A is lost because B did not see it.

This is a very common issue in parallel programming. To prevent it, functions
need to be atomically consistent (aka linearisable), meaning they must have a
linearisation point at which they appear to occur instantly. Such function can

> be written with different technics, including:

Reviewer 6: typo: techniques (and maybe "Such function**s**")

Reviewer 6: Sorry for the "typo" reviews, I missed the warning not to report those.. But otherwise the new readme is a great read, well done! :)

Author: That's alright, I should have removed it  ! Thanks for the review !

function is not linerisable. We could make it atomic by using mutex, but then we
loose the progress properties of the composed functions.

#### Extending Data Structures

Note that in the case above, we transfer from and to a queue of the same
`int Saturn.Queue.t` type. It is most likely possible to write a
`val transfer : t -> t -> unit` function with the right properties and add it
directly to `Saturn.Queue` module.

If you think of any such function, that is useful and missing, let's us know by
creating an issue!

#### Composable Parallelism-Safe Data Structures

If you need composable parallelism-safe data structures, you can check
[kcas_data](https://github.com/ocaml-multicore/kcas#programming-with-transactional-data-structures).

# Testing

One of the many difficulties of implementating parallelism-safe data structures
is that in addition to providing the same safety properties as sequental ones,
they may also have to observe some
[liveness properties](https://en.wikipedia.org/wiki/Safety_and_liveness_properties)
as well as additional safety properties specific to concurrent programming, like
deadlock-freedom.

> In additional of the expected safety properties, the main properties we want to

Reviewer 2: ```suggestion
In addition to the expected safety properties, the main properties we want to
```

test for are:

- linearisability
- lock-freedom for all the lock-free data structures
- no potentially harmful data races

Here is a list of the tools we use to ensure them:

- _safety_ : unitary tests and `qcheck` tests check semantics and expected
  behaviors with one and more domains.
- _safety and liveness_ : `STM` tests check _linearisability_ for two domains
  (see
  [`multicoretests` library](https://github.com/ocaml-multicore/multicoretests)).
- _liveness_ : `dscheck` checks _non-blocking_ property for as many domains as
  wanted (for two domains most of the time). See
  [dscheck](https://github.com/ocaml-multicore/dscheck).
- _safety_ : no data race with
  [tsan](https://github.com/ocaml-multicore/ocaml-tsan)

See [test/README.md](test/README.md) for more details.

> # Benchmarks

Reviewer 2: Not able to comment below: "There are a number of bechmarks".
